File: sysmayal7/sysmayal2/report/distribution_analytics_report/distribution_analytics_report.js
/**
 * Distribution Analytics Report - Client Side Configuration
 */

frappe.query_reports["Distribution Analytics Report"] = {
    "filters": [
        {
            "fieldname": "country",
            "label": __("Country"),
            "fieldtype": "Link",
            "options": "Country",
            "width": "100px"
        },
        {
            "fieldname": "organization_type",
            "label": __("Organization Type"),
            "fieldtype": "Select",
            "options": "\nDistributor\nRetailer\nSupplier\nManufacturer\nWholesaler\nAgent\nConsultant\nOther",
            "width": "120px"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nActive\nInactive\nPending\nSuspended\nTerminated",
            "width": "100px"
        },
        {
            "fieldname": "territory",
            "label": __("Territory"),
            "fieldtype": "Link",
            "options": "Territory",
            "width": "100px"
        },
        {
            "fieldname": "regulatory_status",
            "label": __("Regulatory Status"),
            "fieldtype": "Select",
            "options": "\nCompliant\nPending Review\nNon-Compliant\nExpired\nNot Applicable",
            "width": "120px"
        },
        {
            "fieldname": "min_revenue",
            "label": __("Min Revenue"),
            "fieldtype": "Currency",
            "width": "100px"
        },
        {
            "fieldname": "max_revenue",
            "label": __("Max Revenue"),
            "fieldtype": "Currency",
            "width": "100px"
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        // Color coding for status
        if (column.fieldname == "status") {
            if (value == "Active") {
                value = `<span style="color: #28a745; font-weight: bold;">${value}</span>`;
            } else if (value == "Inactive" || value == "Terminated") {
                value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
            } else if (value == "Pending" || value == "Suspended") {
                value = `<span style="color: #ffc107; font-weight: bold;">${value}</span>`;
            }
        }

        // Color coding for regulatory status
        if (column.fieldname == "regulatory_status") {
            if (value == "Compliant") {
                value = `<span style="color: #28a745; font-weight: bold;">${value}</span>`;
            } else if (value == "Non-Compliant" || value == "Expired") {
                value = `<span style="color: #dc3545; font-weight: bold;">${value}</span>`;
            } else if (value == "Pending Review") {
                value = `<span style="color: #ffc107; font-weight: bold;">${value}</span>`;
            }
        }

        // Highlight expiring agreements
        if (column.fieldname == "days_to_expiry") {
            if (data && data.days_to_expiry !== null) {
                if (data.days_to_expiry < 0) {
                    value = `<span style="color: #dc3545; font-weight: bold;">${value} (Expired)</span>`;
                } else if (data.days_to_expiry <= 30) {
                    value = `<span style="color: #fd7e14; font-weight: bold;">${value}</span>`;
                } else if (data.days_to_expiry <= 90) {
                    value = `<span style="color: #ffc107;">${value}</span>`;
                }
            }
        }

        // Format revenue with indicators
        if (column.fieldname == "annual_revenue") {
            if (data && data.annual_revenue) {
                const revenue = data.annual_revenue;
                let indicator = "";
                
                if (revenue >= 10000000) { // $10M+
                    indicator = ' <span style="color: #28a745;">●</span>';
                } else if (revenue >= 1000000) { // $1M+
                    indicator = ' <span style="color: #ffc107;">●</span>';
                } else if (revenue >= 100000) { // $100K+
                    indicator = ' <span style="color: #17a2b8;">●</span>';
                }
                
                value = value + indicator;
            }
        }

        // Contact count with visual indicator
        if (column.fieldname == "contact_count") {
            if (data && data.contact_count !== null) {
                const count = data.contact_count;
                let color = "#6c757d";
                
                if (count >= 10) {
                    color = "#28a745";
                } else if (count >= 5) {
                    color = "#ffc107";
                } else if (count >= 1) {
                    color = "#17a2b8";
                } else {
                    color = "#dc3545";
                }
                
                value = `<span style="color: ${color}; font-weight: bold;">${count}</span>`;
            }
        }

        return value;
    },

    "onload": function(report) {
        // Add custom buttons
        report.page.add_inner_button(__("Distribution Summary"), function() {
            show_distribution_summary(report);
        });

        report.page.add_inner_button(__("Performance Metrics"), function() {
            show_performance_metrics(report);
        });

        report.page.add_inner_button(__("Contact Analytics"), function() {
            show_contact_analytics(report);
        });

        report.page.add_inner_button(__("Export Analysis"), function() {
            export_distribution_analysis(report);
        });
    }
};

function show_distribution_summary(report) {
    const filters = report.get_values();
    
    frappe.call({
        method: "sysmayal.sysmayal_module.report.distribution_analytics_report.distribution_analytics_report.get_distribution_summary",
        args: { filters: filters },
        callback: function(r) {
            if (r.message) {
                display_summary_dialog(r.message);
            }
        }
    });
}

function show_performance_metrics(report) {
    const filters = report.get_values();
    
    frappe.call({
        method: "sysmayal.sysmayal_module.report.distribution_analytics_report.distribution_analytics_report.get_performance_metrics",
        args: { filters: filters },
        callback: function(r) {
            if (r.message) {
                display_performance_dialog(r.message);
            }
        }
    });
}

function show_contact_analytics(report) {
    const filters = report.get_values();
    
    frappe.call({
        method: "sysmayal.sysmayal_module.report.distribution_analytics_report.distribution_analytics_report.get_contact_analytics",
        args: { filters: filters },
        callback: function(r) {
            if (r.message) {
                display_contact_analytics_dialog(r.message);
            }
        }
    });
}

function display_summary_dialog(data) {
    const summary = data.summary;
    const geographic = data.geographic_distribution;
    const compliance = data.compliance_distribution;
    const expiring = data.expiring_agreements;

    let html = `
        <div class="row">
            <div class="col-md-6">
                <h5>Network Overview</h5>
                <table class="table table-bordered">
                    <tr><td>Total Organizations</td><td><strong>${summary.total_organizations}</strong></td></tr>
                    <tr><td>Active Organizations</td><td><strong>${summary.active_organizations}</strong></td></tr>
                    <tr><td>Countries Covered</td><td><strong>${summary.countries_covered}</strong></td></tr>
                    <tr><td>Territories Covered</td><td><strong>${summary.territories_covered}</strong></td></tr>
                </table>
                
                <h6>Organization Types</h6>
                <table class="table table-bordered">
                    <tr><td>Distributors</td><td>${summary.distributors}</td></tr>
                    <tr><td>Retailers</td><td>${summary.retailers}</td></tr>
                    <tr><td>Suppliers</td><td>${summary.suppliers}</td></tr>
                    <tr><td>Manufacturers</td><td>${summary.manufacturers}</td></tr>
                </table>
            </div>
            
            <div class="col-md-6">
                <h5>Financial Overview</h5>
                <table class="table table-bordered">
                    <tr><td>Total Revenue</td><td><strong>${format_currency(summary.total_revenue)}</strong></td></tr>
                    <tr><td>Average Revenue</td><td><strong>${format_currency(summary.avg_revenue)}</strong></td></tr>
                    <tr><td>Total Employees</td><td><strong>${summary.total_employees || 0}</strong></td></tr>
                    <tr><td>Average Employees</td><td><strong>${Math.round(summary.avg_employees || 0)}</strong></td></tr>
                </table>
                
                <h6>Compliance Status</h6>
                <table class="table table-bordered">
    `;
    
    compliance.forEach(comp => {
        html += `<tr><td>${comp.regulatory_status || 'Not Set'}</td><td>${comp.count} (${comp.percentage}%)</td></tr>`;
    });
    
    html += `
                </table>
            </div>
        </div>
    `;
    
    if (expiring.length > 0) {
        html += `
            <div class="row">
                <div class="col-md-12">
                    <h5>Expiring Agreements (Next 90 Days)</h5>
                    <table class="table table-bordered">
                        <thead>
                            <tr><th>Organization</th><th>Expiry Date</th><th>Days Left</th></tr>
                        </thead>
                        <tbody>
        `;
        
        expiring.forEach(exp => {
            const color = exp.days_to_expiry <= 30 ? 'text-danger' : 'text-warning';
            html += `<tr><td>${exp.organization_name}</td><td>${exp.agreement_expiry}</td><td class="${color}">${exp.days_to_expiry}</td></tr>`;
        });
        
        html += `
                        </tbody>
                    </table>
                </div>
            </div>
        `;
    }

    frappe.msgprint({
        title: __("Distribution Network Summary"),
        message: html,
        wide: true
    });
}

function display_performance_dialog(data) {
    const performance = data.performance_by_type;
    const growth = data.growth_analysis;

    let html = `
        <div class="row">
            <div class="col-md-12">
                <h5>Performance by Organization Type</h5>
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Type</th>
                            <th>Count</th>
                            <th>Avg Revenue</th>
                            <th>Avg Employees</th>
                            <th>Compliance Rate</th>
                        </tr>
                    </thead>
                    <tbody>
    `;
    
    performance.forEach(perf => {
        html += `
            <tr>
                <td>${perf.organization_type}</td>
                <td>${perf.count}</td>
                <td>${format_currency(perf.avg_revenue)}</td>
                <td>${Math.round(perf.avg_employees || 0)}</td>
                <td>${perf.compliance_rate}%</td>
            </tr>
        `;
    });
    
    html += `
                    </tbody>
                </table>
                
                <h5>Growth Analysis by Age</h5>
                <table class="table table-bordered">
                    <thead>
                        <tr><th>Age Group</th><th>Count</th><th>Average Revenue</th></tr>
                    </thead>
                    <tbody>
    `;
    
    growth.forEach(gr => {
        html += `<tr><td>${gr.age_group}</td><td>${gr.count}</td><td>${format_currency(gr.avg_revenue)}</td></tr>`;
    });
    
    html += `
                    </tbody>
                </table>
            </div>
        </div>
    `;

    frappe.msgprint({
        title: __("Performance Metrics"),
        message: html,
        wide: true
    });
}

function display_contact_analytics_dialog(data) {
    const contacts = data.contact_by_role;
    const communication = data.communication_preferences;

    let html = `
        <div class="row">
            <div class="col-md-6">
                <h5>Contacts by Regulatory Role</h5>
                <table class="table table-bordered">
                    <thead>
                        <tr><th>Role</th><th>Total</th><th>Active</th><th>Organizations</th></tr>
                    </thead>
                    <tbody>
    `;
    
    contacts.forEach(contact => {
        html += `
            <tr>
                <td>${contact.regulatory_role || 'Not Set'}</td>
                <td>${contact.count}</td>
                <td>${contact.active_count}</td>
                <td>${contact.organizations_covered}</td>
            </tr>
        `;
    });
    
    html += `
                    </tbody>
                </table>
            </div>
            
            <div class="col-md-6">
                <h5>Communication Preferences</h5>
                <table class="table table-bordered">
                    <thead>
                        <tr><th>Preference</th><th>Count</th><th>Percentage</th></tr>
                    </thead>
                    <tbody>
    `;
    
    communication.forEach(comm => {
        html += `
            <tr>
                <td>${comm.communication_preference || 'Not Set'}</td>
                <td>${comm.count}</td>
                <td>${comm.percentage}%</td>
            </tr>
        `;
    });
    
    html += `
                    </tbody>
                </table>
            </div>
        </div>
    `;

    frappe.msgprint({
        title: __("Contact Analytics"),
        message: html,
        wide: true
    });
}

function export_distribution_analysis(report) {
    const data = report.data;
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data to export"));
        return;
    }

    // Create comprehensive analysis export
    frappe.msgprint(__("Exporting distribution analysis..."));
    
    // In a real implementation, this would generate a comprehensive Excel report
    // with multiple sheets including summary, detailed data, and charts
}

function format_currency(amount) {
    if (!amount) return '$0';
    return new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 0
    }).format(amount);
}

File: sysmayal7/sysmayal2/report/distribution_analytics_report/distribution_analytics_report.py
"""
Distribution Analytics Report

This report provides comprehensive analytics on distribution network
including organization performance, geographic coverage, and growth trends.
"""

import frappe
from frappe import _
from frappe.utils import nowdate, add_months

def execute(filters=None):
    """Execute the distribution analytics report."""
    
    columns = get_columns()
    data = get_data(filters)
    chart_data = get_chart_data(data)
    
    return columns, data, None, chart_data

def get_columns():
    """Define report columns."""
    
    return [
        {
            "fieldname": "organization_name",
            "label": _("Organization"),
            "fieldtype": "Link",
            "options": "Distribution Organization",
            "width": 200
        },
        {
            "fieldname": "organization_type",
            "label": _("Type"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "country",
            "label": _("Country"),
            "fieldtype": "Link",
            "options": "Country",
            "width": 120
        },
        {
            "fieldname": "territory",
            "label": _("Territory"),
            "fieldtype": "Link",
            "options": "Territory",
            "width": 120
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "contact_count",
            "label": _("Contacts"),
            "fieldtype": "Int",
            "width": 80
        },
        {
            "fieldname": "regulatory_status",
            "label": _("Regulatory Status"),
            "fieldtype": "Data",
            "width": 130
        },
        {
            "fieldname": "annual_revenue",
            "label": _("Annual Revenue"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "employee_count",
            "label": _("Employees"),
            "fieldtype": "Int",
            "width": 90
        },
        {
            "fieldname": "established_year",
            "label": _("Established"),
            "fieldtype": "Int",
            "width": 90
        },
        {
            "fieldname": "agreement_expiry",
            "label": _("Agreement Expiry"),
            "fieldtype": "Date",
            "width": 120
        },
        {
            "fieldname": "days_to_expiry",
            "label": _("Days to Expiry"),
            "fieldtype": "Int",
            "width": 110
        },
        {
            "fieldname": "last_audit_date",
            "label": _("Last Audit"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "next_audit_due",
            "label": _("Next Audit Due"),
            "fieldtype": "Date",
            "width": 110
        },
        {
            "fieldname": "business_focus",
            "label": _("Business Focus"),
            "fieldtype": "Data",
            "width": 150
        }
    ]

def get_data(filters):
    """Get report data based on filters."""
    
    conditions = get_conditions(filters)
    
    query = f"""
        SELECT 
            org.name,
            org.organization_name,
            org.organization_type,
            org.country,
            org.territory,
            org.status,
            org.regulatory_status,
            org.annual_revenue,
            org.employee_count,
            org.established_year,
            org.agreement_expiry,
            org.last_audit_date,
            org.next_audit_due,
            org.business_focus,
            (SELECT COUNT(*) 
             FROM `tabDistribution Contact` dc 
             WHERE dc.organization = org.name 
             AND dc.status = 'Active') as contact_count,
            CASE 
                WHEN org.agreement_expiry IS NOT NULL 
                THEN DATEDIFF(org.agreement_expiry, CURDATE())
                ELSE NULL 
            END as days_to_expiry
        FROM `tabDistribution Organization` org
        WHERE org.docstatus < 2
        {conditions}
        ORDER BY 
            org.country,
            org.organization_type,
            org.organization_name
    """
    
    return frappe.db.sql(query, as_dict=True)

def get_conditions(filters):
    """Build SQL conditions based on filters."""
    
    conditions = ""
    
    if filters:
        if filters.get("country"):
            conditions += f" AND org.country = '{filters['country']}'"
            
        if filters.get("organization_type"):
            conditions += f" AND org.organization_type = '{filters['organization_type']}'"
            
        if filters.get("status"):
            conditions += f" AND org.status = '{filters['status']}'"
            
        if filters.get("territory"):
            conditions += f" AND org.territory = '{filters['territory']}'"
            
        if filters.get("regulatory_status"):
            conditions += f" AND org.regulatory_status = '{filters['regulatory_status']}'"
            
        if filters.get("min_revenue"):
            conditions += f" AND org.annual_revenue >= {filters['min_revenue']}"
            
        if filters.get("max_revenue"):
            conditions += f" AND org.annual_revenue <= {filters['max_revenue']}"
            
    return conditions

def get_chart_data(data):
    """Generate chart data for the report."""
    
    # Organization type distribution
    type_counts = {}
    country_counts = {}
    status_counts = {}
    revenue_by_type = {}
    
    for row in data:
        # Type distribution
        org_type = row.get('organization_type', 'Unknown')
        type_counts[org_type] = type_counts.get(org_type, 0) + 1
        
        # Country distribution
        country = row.get('country', 'Unknown')
        country_counts[country] = country_counts.get(country, 0) + 1
        
        # Status distribution
        status = row.get('status', 'Unknown')
        status_counts[status] = status_counts.get(status, 0) + 1
        
        # Revenue by type
        if org_type not in revenue_by_type:
            revenue_by_type[org_type] = 0
        if row.get('annual_revenue'):
            revenue_by_type[org_type] += row['annual_revenue']
    
    chart_data = {
        "data": {
            "labels": list(type_counts.keys()),
            "datasets": [{
                "name": "Organizations by Type",
                "values": list(type_counts.values())
            }]
        },
        "type": "bar",
        "height": 300,
        "colors": ["#28a745", "#17a2b8", "#ffc107", "#dc3545", "#6f42c1"]
    }
    
    return chart_data

@frappe.whitelist()
def get_distribution_summary(filters=None):
    """Get distribution network summary statistics."""
    
    conditions = get_conditions(filters) if filters else ""
    
    # Basic counts
    summary_query = f"""
        SELECT 
            COUNT(*) as total_organizations,
            COUNT(CASE WHEN status = 'Active' THEN 1 END) as active_organizations,
            COUNT(CASE WHEN organization_type = 'Distributor' THEN 1 END) as distributors,
            COUNT(CASE WHEN organization_type = 'Retailer' THEN 1 END) as retailers,
            COUNT(CASE WHEN organization_type = 'Supplier' THEN 1 END) as suppliers,
            COUNT(CASE WHEN organization_type = 'Manufacturer' THEN 1 END) as manufacturers,
            COUNT(DISTINCT country) as countries_covered,
            COUNT(DISTINCT territory) as territories_covered,
            AVG(annual_revenue) as avg_revenue,
            SUM(annual_revenue) as total_revenue,
            AVG(employee_count) as avg_employees,
            SUM(employee_count) as total_employees
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        {conditions}
    """
    
    summary = frappe.db.sql(summary_query, as_dict=True)[0]
    
    # Geographic distribution
    geo_query = f"""
        SELECT 
            country,
            COUNT(*) as org_count,
            AVG(annual_revenue) as avg_revenue,
            COUNT(CASE WHEN status = 'Active' THEN 1 END) as active_count
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        {conditions}
        GROUP BY country
        ORDER BY org_count DESC
    """
    
    geographic_data = frappe.db.sql(geo_query, as_dict=True)
    
    # Compliance distribution
    compliance_query = f"""
        SELECT 
            regulatory_status,
            COUNT(*) as count,
            ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM `tabDistribution Organization` WHERE docstatus < 2 {conditions}), 1) as percentage
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        {conditions}
        GROUP BY regulatory_status
        ORDER BY count DESC
    """
    
    compliance_data = frappe.db.sql(compliance_query, as_dict=True)
    
    # Expiring agreements
    expiring_query = f"""
        SELECT 
            organization_name,
            agreement_expiry,
            DATEDIFF(agreement_expiry, CURDATE()) as days_to_expiry
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        AND agreement_expiry IS NOT NULL
        AND agreement_expiry BETWEEN CURDATE() AND DATE_ADD(CURDATE(), INTERVAL 90 DAY)
        {conditions}
        ORDER BY agreement_expiry
    """
    
    expiring_agreements = frappe.db.sql(expiring_query, as_dict=True)
    
    return {
        "summary": summary,
        "geographic_distribution": geographic_data,
        "compliance_distribution": compliance_data,
        "expiring_agreements": expiring_agreements
    }

@frappe.whitelist()
def get_performance_metrics(filters=None):
    """Get performance metrics for distribution network."""
    
    conditions = get_conditions(filters) if filters else ""
    
    # Performance by type
    performance_query = f"""
        SELECT 
            organization_type,
            COUNT(*) as count,
            AVG(annual_revenue) as avg_revenue,
            AVG(employee_count) as avg_employees,
            COUNT(CASE WHEN regulatory_status = 'Compliant' THEN 1 END) as compliant_count,
            ROUND(COUNT(CASE WHEN regulatory_status = 'Compliant' THEN 1 END) * 100.0 / COUNT(*), 1) as compliance_rate
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        {conditions}
        GROUP BY organization_type
        ORDER BY count DESC
    """
    
    performance_data = frappe.db.sql(performance_query, as_dict=True)
    
    # Growth analysis (based on established year)
    growth_query = f"""
        SELECT 
            CASE 
                WHEN established_year >= YEAR(CURDATE()) - 2 THEN 'New (0-2 years)'
                WHEN established_year >= YEAR(CURDATE()) - 5 THEN 'Young (3-5 years)'
                WHEN established_year >= YEAR(CURDATE()) - 10 THEN 'Mature (6-10 years)'
                ELSE 'Established (10+ years)'
            END as age_group,
            COUNT(*) as count,
            AVG(annual_revenue) as avg_revenue
        FROM `tabDistribution Organization`
        WHERE docstatus < 2
        AND established_year IS NOT NULL
        {conditions}
        GROUP BY age_group
        ORDER BY 
            CASE age_group
                WHEN 'New (0-2 years)' THEN 1
                WHEN 'Young (3-5 years)' THEN 2
                WHEN 'Mature (6-10 years)' THEN 3
                ELSE 4
            END
    """
    
    growth_data = frappe.db.sql(growth_query, as_dict=True)
    
    return {
        "performance_by_type": performance_data,
        "growth_analysis": growth_data
    }

@frappe.whitelist()
def get_contact_analytics(filters=None):
    """Get contact analytics for distribution network."""
    
    org_conditions = get_conditions(filters) if filters else ""
    
    # Contact distribution by role
    contact_query = f"""
        SELECT 
            dc.regulatory_role,
            COUNT(*) as count,
            COUNT(CASE WHEN dc.status = 'Active' THEN 1 END) as active_count,
            COUNT(DISTINCT dc.organization) as organizations_covered
        FROM `tabDistribution Contact` dc
        JOIN `tabDistribution Organization` org ON dc.organization = org.name
        WHERE dc.docstatus < 2
        AND org.docstatus < 2
        {org_conditions.replace('org.', 'org.')}
        GROUP BY dc.regulatory_role
        ORDER BY count DESC
    """
    
    contact_data = frappe.db.sql(contact_query, as_dict=True)
    
    # Communication preferences
    comm_query = f"""
        SELECT 
            dc.communication_preference,
            COUNT(*) as count,
            ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM `tabDistribution Contact` WHERE docstatus < 2), 1) as percentage
        FROM `tabDistribution Contact` dc
        JOIN `tabDistribution Organization` org ON dc.organization = org.name
        WHERE dc.docstatus < 2
        AND org.docstatus < 2
        {org_conditions.replace('org.', 'org.')}
        GROUP BY dc.communication_preference
        ORDER BY count DESC
    """
    
    communication_data = frappe.db.sql(comm_query, as_dict=True)
    
    return {
        "contact_by_role": contact_data,
        "communication_preferences": communication_data
    }
